### PR TITLE
Fix logical functions (or, and, xor) when we have more than 10 arguments

### PR DESCRIPTION
**Pull request by YiuRULE**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):

Functions `or`, `and` and `xor` return correct values when we send more than 10 arguments

Detailed description (optional):

Related to https://github.com/ClickHouse/ClickHouse/issues/8469

ClickHouse possibly return false result for the functions `or`, `and` and `xor` when we use more than 10 arguments.

For example, if we create this table, with this data

```
CREATE TABLE foo (
    `id` UInt32,
    `status` Nullable(UInt32)
)  ENGINE = MergeTree()
   PARTITION BY id % 5
   PRIMARY KEY (id)
   ORDER BY (id, assumeNotNull(status));

INSERT INTO foo SELECT number, nullIf(number % 3, 1) from system.numbers LIMIT 1000000;
```

And we use this `SELECT` query

```
SELECT count()
FROM foo
WHERE ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1))
      OR ((status >= 1) AND (status <= 1));
```

It return `333333`, where his expected behaviour should be `0`.

It was due than the current filtering was translated to something like `or(and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)), and(moreOrEqual(1), lessOrEqual(1)))`

The main problem is the current implementation have an hard limit of 10 arguments, due to some meta-programming magic when we do the actual operation, defined there

https://github.com/ClickHouse/ClickHouse/blob/master/dbms/src/Functions/FunctionsLogical.cpp#L281-L283

So when we send more than 10 arguments, we could have some undefined behavior as some columns has not been computed.

The proposed solution on this PR is to create a `Runtime` operations when we cannot handle the operation on the compile-time.

So when we have less or 10 arguments, we rely on the compile-time behavior for optimization purpose, otherwise we call the runtime one.

### CODE REVIEW

**Comment by YiuRULE** [comment]
I rebase the branch from master.
Small note: The CI tests who fail are not related to this PR. (reproducible to master)

**Comment by YiuRULE** [comment]
Related to https://github.com/ClickHouse/ClickHouse/issues/8469 btw

**Comment by YiuRULE** [comment]
@Akazz Hi Akazz, did you had time to check a bit about this PR ? does the solution seem ok ?

**Comment by Akazz** [comment]
> @Akazz Hi Akazz, did you had time to check a bit about this PR ? does the solution seem ok ?

I've identified the problem and currently am working on a solution. Please wait for a soon update, I'll post all of the details

File: dbms/src/Functions/FunctionsLogical.cpp [hunk 1]
```diff
@@ -275,11 +275,39 @@ class AssociativeGenericApplierImpl<Op, 1>
...
+{
+    using ResultValueType = typename Op::ResultType;
+
+public:
+    AssociativeGenericRuntimeApplierImpl(const ColumnRawPtrs & _in)
+        : in{_in} {}
+
+    inline ResultValueType apply(const size_t i, const size_t column_counter) const
+    {
+        const auto a = ValueGetterBuilder::build(in[in.size() - column_counter])(i);
```
Comment (Akazz) [inline dbms/src/Functions/FunctionsLogical.cpp]: Unfortunately this line will, for one, ruin any attempts of vectorized execution optimizations

**Review by Akazz — COMMENTED** [review]
The root cause of the problem is that the algorithm does not use correct internal representation of intermediate result when it comes to calculating in multiple batches. Also, I am sure you would agree with me on that changes which fix this sort of bug utimately require tests! I suggest we close this PR and move over to #8718

**Pull request closed without merging**